########################################################
###   CenTFinder                                     ###
###   Class definitions                              ###
###   version 21032018                               ###
###   Author: Jonathan De Smedt                      ###
########################################################

# Set old classes

setOldClass("hclust")

# Class for CenTFinder

#' An S4 class to represent a CenTFinder analysis
#'
#' @export CenTFinder
#'
#' @slot data A gene expression matrix with gene symbols in the rows and samples in the columns.
#' @slot sample_annotation Annotation data for each sample (including cell type, array, and accession code.)
#' @slot platforms Platform usage information.
#' @slot probe_annotations List of additionally provided probe annotations.
#' @slot gene_cluster_annotation Gene-to-cluster annotation.
#' @slot DE A list of differential expression analyses, each element corresponding to one of the cell types.
#' @slot additional_DE Additional differential expression data for the cell type of interest.
#' @slot plots Plots generated by WGCNA analysis.
#' @slot power The power used for WGCNA soft-tresholding.
#' @slot TOM WGCNA topological overlap matrix.
#' @slot tree Gene tree generated by hierarchical clustering.
#' @slot module_colors Module colour of each gene.
#' @slot kMEs Module eigengenes for each gene and each module.
#' @slot GO List of Gene Ontology analysis results of each module.
#' @slot KEGG List of KEGG enrichment analysis results of each module.
#' @slot motif_enrichment Results of the RCisTarget analyses.
#' @slot integration List containing an integrative summary of the different analyses, module and regulon scores, and the regulons formatted for Cytoscape visualisation.

CenTFinder<-setClass("CenTFinder",
         slots = c(
           data = "matrix",
           sample_annotation = "data.frame",
           platforms = "data.frame",
           probe_annotations = "list",
           gene_cluster_annotation = "data.frame",
           DE = "list",
           additional_DE = "list",
           plots = "list",
           power = "numeric",
           TOM = "matrix",
           tree = "hclust",
           module_colors = "character",
           kMEs = "data.frame",
           GO = "list",
           KEGG = "list",
           motif_enrichment = "data.frame",
           integration = "list"
         ))







